pkg/controllers/externaloidc/externaloidc_controller_test.go

```diff
@@ -0,0 +1,964 @@
+			expectError: false,
+		},
+		{
+			name:                 "auth type IntegratedOAuth with auth configmap",
+			existingAuthConfigCM: &baseAuthConfigCM,
+			caBundleConfigMap:    &baseCABundleConfigMap,
+			auth:                 newAuthWithSpec(configv1.AuthenticationSpec{Type: configv1.AuthenticationTypeIntegratedOAuth}),
+			expectEvents:         true,
+			expectError:          false,
+		},
+		{
+			name:             "auth type OIDC but config map lister fails",
+			configMapIndexer: cache.Indexer(&everFailingIndexer{}),
+			expectEvents:     false,
+			expectError:      true,
+		},
+		{
+			name:                 "auth type OIDC config same as existing",
+			existingAuthConfigCM: authConfigCMWithIssuerURL(&baseAuthConfigCM, testServer.URL),
+			auth: authWithUpdates(baseAuthResource, []func(auth *configv1.Authentication){
+				func(auth *configv1.Authentication) {
+					auth.Spec.OIDCProviders[0].Issuer.URL = testServer.URL
+				},
+			}),
+			expectedAuthConfigJSON: strings.ReplaceAll(baseAuthConfigJSON, "$URL", testServer.URL),
+			caBundleConfigMap:      &baseCABundleConfigMap,
+			expectEvents:           false,
+			expectError:            false,
+		},
+		{
+			name: "auth type OIDC error while applying config",
+			cmApplyReaction: func(action k8stesting.Action) (bool, runtime.Object, error) {
+				return true, nil, fmt.Errorf("error")
+			},
+			existingAuthConfigCM: &baseAuthConfigCM,
+			auth: authWithUpdates(baseAuthResource, []func(auth *configv1.Authentication){
+				func(auth *configv1.Authentication) {
+					auth.Spec.OIDCProviders[0].Issuer.Audiences = []configv1.TokenAudience{"my-test-aud", "yet-another-aud"}
+				},
+			}),
+			expectEvents: false,
+			expectError:  true,
+		},
+		{
+			name:                 "auth type OIDC apply config",
+			caBundleConfigMap:    &baseCABundleConfigMap,
+			existingAuthConfigCM: authConfigCMWithIssuerURL(&baseAuthConfigCM, testServer.URL),
+			auth: authWithUpdates(baseAuthResource, []func(auth *configv1.Authentication){
+				func(auth *configv1.Authentication) {
+					auth.Spec.OIDCProviders[0].Issuer.URL = testServer.URL
+					auth.Spec.OIDCProviders[0].Issuer.Audiences = []configv1.TokenAudience{"my-test-aud", "yet-another-aud"}
+				},
+			}),
+			expectedAuthConfigJSON: func() string {
+				str := strings.ReplaceAll(baseAuthConfigJSON, "$URL", testServer.URL)
+				str = strings.ReplaceAll(str, "another-aud", "yet-another-aud")
+				return str
+			}(),
+			expectEvents: true,
+			expectError:  false,
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			objects := []runtime.Object{}
+
+			authIndexer := cache.NewIndexer(func(obj interface{}) (string, error) {
+				return "cluster", nil
+			}, cache.Indexers{})
+
+			if tt.configMapIndexer == nil {
+				tt.configMapIndexer = cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{})
+			}
+
+			if tt.auth != nil {
+				authIndexer.Add(tt.auth)
```

Comment:
Nit: I've never seen the use of indexers for mocking out test data before, and while it makes sense, I wonder if creating mocks for the lister/getter interfaces could be easier to reason about and/or simplify the complexity of the test logic a bit?